File: android/app/src/main/java/com/soohoon/influe/data/model/SponsorshipDTO.kt
package com.soohoon.influe.data.model

import kotlinx.datetime.Instant
import kotlinx.datetime.Clock
import kotlinx.datetime.TimeZone
import kotlinx.datetime.daysUntil
import kotlinx.datetime.toLocalDateTime
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

enum class SponsorshipStatus(val rawValue: String, val displayName: String) {
    PRE_SUBMIT("preSubmit", "제출 전"),
    UNDER_REVIEW("underReview", "검수중"),
    SUBMITTED("submitted", "제출 완료"),
    PENDING_SETTLEMENT("pendingSettlement", "정산 대기"),
    COMPLETED("completed", "완료");

    companion object {
        fun fromRawValue(raw: String): SponsorshipStatus =
            entries.find { it.rawValue == raw } ?: PRE_SUBMIT
    }
}

@Serializable
data class SponsorshipDTO(
    val id: String,
    @SerialName("workspace_id") val workspaceId: String,
    @SerialName("brand_name") val brandName: String,
    @SerialName("product_name") val productName: String = "",
    val details: String = "",
    val amount: Double = 0.0,
    @SerialName("start_date") val startDate: String, // ISO 8601
    @SerialName("end_date") val endDate: String,
    val status: String = "preSubmit",
    @SerialName("created_by") val createdBy: String? = null,
    @SerialName("created_at") val createdAt: String = "",
    @SerialName("updated_at") val updatedAt: String = ""
) {
    val sponsorshipStatus: SponsorshipStatus
        get() = SponsorshipStatus.fromRawValue(status)

    val daysRemaining: Int
        get() = try {
            val end = Instant.parse(endDate)
            val now = Clock.System.now()
            now.daysUntil(end, TimeZone.currentSystemDefault())
        } catch (_: Exception) { 0 }

    val isExpired: Boolean get() = daysRemaining < 0

    val isExpiringSoon: Boolean get() = daysRemaining in 0..3
}

@Serializable
data class SponsorshipInsert(
    @SerialName("workspace_id") val workspaceId: String,
    @SerialName("created_by") val createdBy: String,
    @SerialName("brand_name") val brandName: String,
    @SerialName("product_name") val productName: String = "",
    val details: String = "",
    val amount: Double = 0.0,
    @SerialName("start_date") val startDate: String,
    @SerialName("end_date") val endDate: String,
    val status: String = "preSubmit"
)

File: CreatorNote/Supabase/DTOs/SponsorshipDTO.swift
import Foundation

struct SponsorshipDTO: Codable, Identifiable, Sendable {
    let id: UUID
    var workspaceId: UUID
    var brandName: String
    var productName: String
    var details: String
    var amount: Double
    var startDate: Date
    var endDate: Date
    var status: String
    var isPinned: Bool
    var createdBy: UUID?
    var updatedBy: UUID?
    var createdAt: Date
    var updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id, details, amount, status
        case workspaceId = "workspace_id"
        case brandName = "brand_name"
        case productName = "product_name"
        case startDate = "start_date"
        case endDate = "end_date"
        case isPinned = "is_pinned"
        case createdBy = "created_by"
        case updatedBy = "updated_by"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    var daysRemaining: Int {
        Calendar.current.dateComponents([.day], from: .now, to: endDate).day ?? 0
    }

    var isExpired: Bool {
        endDate < .now
    }

    var isExpiringSoon: Bool {
        daysRemaining <= 3 && daysRemaining >= 0
    }

    var sponsorshipStatus: SponsorshipStatus {
        SponsorshipStatus(rawValue: status) ?? .preSubmit
    }
}
